zerver/lib/users.py

```diff
@@ -342,8 +342,59 @@ def access_user_by_email(
     allow_bots: bool = False,
     for_admin: bool,
 ) -> UserProfile:
+    """
+    Fetch a user by email address. Endpoints using this function can be queried either with:
+    1) The real email address of the intended user, if the requester believes that user exists and allows
+       their email address to be visible to the requester.
+    2) The dummy email address (of the approximate shape 'user<id>@<dummy email domain>') of the intended
+       user. We detect when the format of the provided email address matches the format of our dummy
+       email addresses and extract the user id from it for a regular id-based lookup.
+
+       In particular, this mode is kept around for backwards compatibility with the old behavior of the
+       `GET /users/{email}` endpoint, which required use of the dummy email address for lookups of any user
+       which didn't have EMAIL_ADDRESS_VISIBILITY_EVERYONE set, regardless of how the actual
+       email_address_visibility setting related to the role of the requester.
+
+       Note: If the realm.host value changes (e.g. due to the server moving to a new
+       domain), the required dummy email values passed here will need to be updated
+       accordingly to match the new value. This deviates from the original API behavior,
+       where the lookups were supposed to match the UserProfile.email value, which was
+       **not** updated for existing users even if the server moved domains. See
+       get_fake_email_domain for details of how the email domain for dummy email addresses
+       is determined.
+
+    The purpose of this is to be used at API endpoints that allow selecting the target user by
+    delivery_email, while preventing the endpoint from leaking information about user emails.
+    """
```

Comment:
this is the docstring to review for understanding the intended behavior I'm implementing here.

Reply:
Of course the alternative is that (2) can instead be an `.email`-based lookup like the original behavior, but I feel the `detect the dummy email format and extract the id` approach is much nicer.

Reply:
Yeah I think that's nicer and I doubt that it'll break material code. I added a deprecation note for the old style as well.